### PR TITLE
Cale/octomerge

### PR DESCRIPTION
**Pull request by cpennington**

### CODE REVIEW

File: tubular/git_repo.py [hunk 1]
```diff
@@ -37,36 +38,95 @@ def extract_repo_name(repo_url):
...
+        """
+        self.checkout_branch(target_branch)
+        self.repo.git.merge(source_branch, ff_only=ff_only)
+        merge_sha = self.repo.git.rev_parse('HEAD')
+        return merge_sha
+
+    def add_remote(self, remote_name, remote_url):
+        """
+        Add a remote named ``remote_name`` pointing to ``remote_url``
+        to the repo at ``repo_path``.
```
Comment (doctoryes) [inline tubular/git_repo.py]: `repo_path` isn't a param here?

File: tubular/git_repo.py [hunk 2]
```diff
@@ -37,36 +38,95 @@ def extract_repo_name(repo_url):
...
+        """
+        remote = self.repo.create_remote(remote_name, remote_url)
+        remote.fetch()
+
+    def octopus_merge(self, base_branch, commitishes):
+        """
+        Merge all ``commitishes`` into ``base_branch``, inside ``repo_path``.
```
Comment (doctoryes) [inline tubular/git_repo.py]: `repo_path` isn't a param here?

File: tubular/git_repo.py [hunk 3]
```diff
@@ -37,36 +38,95 @@ def extract_repo_name(repo_url):
...
+
+    def add_remote(self, remote_name, remote_url):
+        """
+        Add a remote named ``remote_name`` pointing to ``remote_url``
+        to the repo at ``repo_path``.
+        """
+        remote = self.repo.create_remote(remote_name, remote_url)
+        remote.fetch()
+
+    def octopus_merge(self, base_branch, commitishes):
```
Comment (doctoryes) [inline tubular/git_repo.py:104]: Great name!

File: tubular/git_repo.py [hunk 4]
```diff
@@ -37,36 +38,95 @@ def extract_repo_name(repo_url):
...
+            repo_url (str): The full url of the repo to clone. The last part of this
+                url will be used as the repository directory name.
+            branch (str): The branch to clone from
+        """
+        repo = Repo.clone_from(
+            repo_url,
+            to_path=extract_repo_name(repo_url),
+            branch=branch
+        )
+        return cls(repo)
```
Comment (doctoryes) [inline tubular/git_repo.py:64]: Fancy. What are the advantages of implementing this code vs. accepting `repo_url` & `branch` in an `__init__` method and cloning there?

File: tubular/git_repo.py [hunk 5]
```diff
@@ -37,36 +38,95 @@ def extract_repo_name(repo_url):
...
+        Merge all ``commitishes`` into ``base_branch``, inside ``repo_path``.
+        """
+        self.checkout_branch(base_branch)
+        self.repo.git.merge(*commitishes)
+        return self.repo.head.commit.hexsha
+
+    def force_branch_to(self, branch, commitish, remote=None):
+        """
+        Reset branch to commitish.
+        """
```
Comment (doctoryes) [inline tubular/git_repo.py]: The docstring should explain what the `remote` param does here.

File: tubular/scripts/merge_approved_prs.py [hunk 6]
```diff
@@ -3,34 +3,35 @@
...
+    Yield all PRs in ``target_repo`` which meet the following criteria:
         * have been approved
         * have not been closed
-        * have a base branch of ``private_base_branch``
-        * have not been merged to ``public_base_branch`` in ``public_repo``
+        * have a base branch of ``target_base_branch``
+        * have not been merged to ``source_base_branch`` in ``source_repo``
 
     Arguments:
         token: A github access token
```
Comment (doctoryes) [inline tubular/scripts/merge_approved_prs.py]: `token` was removed.

File: tubular/scripts/merge_approved_prs.py [hunk 7]
```diff
@@ -41,41 +42,86 @@ def find_approved_prs(token, private_repo, public_repo, private_base_branch, pub
...
+        local_repo.add_remote('source', source_github_repo.github_repo.ssh_url)
+        local_repo.force_branch_to(target_branch, source_branch, remote='source')
+
+        approved_prs = list(find_approved_prs(
+            target_github_repo, source_github_repo, target_base_branch, source_base_branch
+        ))
+        logging.info("Merging the following prs into {}:\n{}".format(
+            target_branch,
+            "\n".join(
+                "    {pr.head.repo.owner.name}/{pr.head.repo.name}#{pr.number}".format(pr=pr)
```
Comment (doctoryes) [inline tubular/scripts/merge_approved_prs.py]: Perhaps output the PR URLs here instead?

File: tubular/git_repo.py (hunk 4 above)
Comment (cpennington) [reply tubular/git_repo.py:64]: Clearer separation of concerns. It allows you to work with a local git repo that has already been cloned by passing it as an instantiated `Repo` object to the constructor.

File: tubular/git_repo.py (hunk 1 above)
Comment (cpennington) [reply tubular/git_repo.py]: Oops.

File: tubular/git_repo.py (hunk 2 above)
Comment (cpennington) [reply tubular/git_repo.py]: Oops.

File: tubular/git_repo.py (hunk 5 above)
Comment (cpennington) [reply tubular/git_repo.py]: Agreed.

File: tubular/scripts/merge_approved_prs.py (hunk 6 above)
Comment (cpennington) [reply tubular/scripts/merge_approved_prs.py]: Oops.

File: tubular/scripts/merge_approved_prs.py (hunk 7 above)
Comment (cpennington) [reply tubular/scripts/merge_approved_prs.py]: Yeah, will do.

**Review by doctoryes — APPROVED** [review]

**Pull request merged**